Improve ArrayList implementation

maksbotan opened the pull request:
Current implementation of `ArrayList` has several problems:

- It grows the vector by factor of 2 when there is not enough space for elements, while it's nowadays common to use factors strictly less then two in vector implementations.
- It only supports `push` operation, not allowing to extend the vector by many elements at once.
- Most importantly, it doesn't expose the API to use the underlying mutable vector, instead users have to read `MutVar`s themselves.
- All functions are generic in monad and vector and lack inline pragmas.
- And it's not documented at all.

I've rewritten the implementation to fix all the points above, using growth factor of `1.5`.

Additionally, I achieved compatibility with newer `primitive` package, which will be included with LTS-15 with GHC-8.8.

maksbotan committed:
Relax depends, adapt to primitive-0.7

primitive-0.7.0.0 dropped Addr type in favor of "Ptr Word8". This
commits adds support for it via CPP.

maksbotan committed:
Improve implementation of ArrayList

Introduce auto-growing vector with over-allocation of memory to save
memory moves, as in C++, Python or Java.

External API of ArrayList does not change.

andrewthad commented:
This looks fine to me. Also, I've given you commit bit, so you can just merge it whenever you feel that it's ready.

maksbotan commented:
Thanks @andrewthad !

Would you mind a version bump as well?

andrewthad commented:
Yeah, feel free to bump the version.

maksbotan committed:
Add SPECIALIZE pragmas to push and grow

These pragmas allow GHC to avoid passing dictionaries for PrimMonad and
MVector instances at use-site, resulting in measurable performance
improvement.

maksbotan committed:
Update stack.yaml to Stack 2 and LTS-14.25

maksbotan committed:
version 0.5.1: ArrayList improvements

maksbotan commented:
Well, I've merged it.
Can you push it to Hackage as well (version 0.5.1)? I will then include it into stackage nightly.
